### ISSUE TITLE
[Featuer Request] Linter

### ISSUE THREAD

**Issue by n-gao** [typst/typst#151]
Thanks a lot for the great alternative to LaTeX! To ease adoption and use, an extension to VSCode similar to latex-workshop would be great! But since the setup is relatively simple compared to LaTeX, a simple linter would already be a great addition to the tooling.

**Comment by clbarnes** [comment]
This is probably a duplicate of https://github.com/typst/typst/issues/118

**Comment by reknih** [comment]
True, there's an overlap between this and #118. We'll consider that issue to be about syntax highlighting grammars and a LSP while we will use this issue to track the feature request of a linter providing hints on patterns and anti-patterns in Typst code.

**Comment by Sinthoras7** [comment]
Shouldn't this issue be in the issue tracker of the IDE not the compiler and therefore be closed here?

**Comment by YDX-2147483647** [comment]
For reference, [the tinymist-lint crate](https://github.com/Myriad-Dreamin/tinymist/tree/main/crates/tinymist-lint) provides basic linting.

Example:

- No bug-prone show/set rules: Use `set text(red) if condition` rather than `if condition { set text(red) }`.
- Warn when using variable font, because of #185.

[![Image](https://github.com/user-attachments/assets/31739da2-7913-4b30-98cc-167053792951)](https://github.com/Myriad-Dreamin/tinymist/pull/1634)

A demo video is available in https://github.com/Myriad-Dreamin/tinymist/releases/tag/v0.13.12 (scroll down to _Linting (New)_).

---

Another way is to create your own lint rules by leveraging [ast-grep](https://ast-grep.github.io/) and https://github.com/uben0/tree-sitter-typst. More info can be found in [my post on Typst Forum](https://forum.typst.app/t/adding-typst-support-to-ast-grep-and-some-package-statistics/5161).

Example: Use show-set in favor of show-function.

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/d4b75102-86ee-4c07-bf06-6351d94179db" />
